docs/DONOTMERGE/pr-402.md

```diff
@@ -0,0 +1,114 @@
+- The client has access to the per-block ledger events, or at least the ones known to be desirable (eg calculated rewards and calculated stake distributions).
+- The client need not hold its own copy of the ledger state in memory.
+  (UTxO HD is related, but does not necessarily completely remove this concern.)
+
+## Solution Space
+
+Today's primary options are as follows.
+
+- Use ChainSync to follow the chain, and maintain a second copy of the ledger state.
+    - REQUIREMENT VIOLATION: This can only be implemented in Haskell.
+    - REQUIREMENT VIOLATION: This violates the desideratum regarding holding second copy of the ledger state in memory.
+      As of UTxO HD, the major portions of those two ledger states would be on disk instead of memory, but there'd still need to be two copies (later development could _maybe_ allow sharing the disk part).
+    - To handle rollbacks, the client would also need to maintain its own copy of Consensus's `LedgerDB` abstraction.
+    - (The [`foldBlocks`](https://input-output-hk.github.io/cardano-node/cardano-api/lib/Cardano-Api-LedgerState.html#v:foldBlocks) approach does this, but only for the immutable chain.)
+
+- Piggy-back on a `db-sync` instance.
+    - REQUIREMENT VIOLATION: This violates the desideratum regarding holding second copy of the ledger state in memory.
+    - @kderme remarked that he and @jasagredo charted a path towards using UTxO HD to ameliorate the memory cost, as sketched in the ChainSync option above.
+    - @KtorZ remarked that `db-sync` currently requires hundreds of gigabytes of disk, so that is also client app burden for this approach.
+
+- Rely on PR #402's `rnHandleLedgerEvent` et al.
+    - REQUIREMENT VIOLATION: This violates the desideratum about becoming synced with a node.
+      The only workaround at the moment is to restart the node and force it to re-apply whichver historical blocks the client needs to process.
+      Moreover, currently, the only way to force the node to do so is to delete certains parts of its on-disk state!
+    - To handle rollbacks, the client would also need to maintain something akin to Consensus's `LedgerDB` abstraction, but more lightweight since its elements need not be entire Cardano ledger states.
+    - @kderme also remarked that `db-sync` only uses the ledger events for so much, and so still relies on the ledger state itself for the rest; this suggests that the blocks and their ledger events alone might not suffice for clients.
+
+The envisioned longer-term solutions are as follows.
+
+- Enrich the node to permanently store ledger events alongside blocks.
+    - Then the ChainSync approach above could simply serve ledger events alongside the blocks.
+    - This solution would very explicitly violate the design principles of the node, since this is non-trivial logic in the node that is absolutely not required for the core functionality of the blockchain itself.
```

Comment:
Can't we make the same argument with the state-query mini protocol? Seems to me like something that is also not required for the core functionality of the blockchain.

Reply:
That is a good point, I think the word "non-trivial" in 

> this is non-trivial logic in the node that is absolutely not required for the core functionality of the blockchain itself

is doing the heavy lifting here, with the main difference between on-disk storage of ledger events and the local state query protocol being that the latter does not require any additional state/storage beyond what the node already maintains for other reasons. 

Definitely makes sense to expand a bit on this :+1:, additional considerations might include: 

 - Maybe certain information available via the state query protocol are crucial to get directly from a node for SPOs (eg the leadership schedule indirectly).
 - Make "non-trivial logic" more precise, eg the state query protocol still has *some* maintenance costs, like ensuring backwards-compatible serialization, and hence is not "trivial", but still "easy enough" to be acceptable into the Node. But maybe it could be even simpler?